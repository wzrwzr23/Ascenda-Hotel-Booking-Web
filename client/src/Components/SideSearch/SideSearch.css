h2 {
    color: white;
    margin: 10px 10px 10px 10px;
}
span {
    color: white;
    margin: 10px 10px 10px 10px;
}

.dest-container {
    display: flex;
    justify-content: center;
    margin-top: 20px;
    padding-bottom: 20px;
}

.destWrapper{
    width: 400px;
    display: flex;
    gap: 20px;
}

/* #usersearch input {
    width: 100%;
} */

.searchBtn{
    display: flex;
    justify-content: center;
    background-color: #bc9a39;  
    color: #e4d2a0;
    text-decoration: none;
    padding:6px;
}

.searchItems{
    flex: 1;
    background-color: #febb02;
    padding: 10px;
    border-radius: 10px;
    position: sticky;
    top: 10px;
    height: max-content;
    flex-direction: column;
}

input {
    width: 40%;
    margin: 10px 10px 10px 10px;
    height: 30px;
    padding-left: 10px;
}
.usersearch input{
    width: 90%;
}

button {
    width: 100%;
    height: 30px;
    margin-top: 10px;
    cursor: pointer;
    border: none;
    color: white;
    background-color: #bc9a39;
}
button:hover {
    background-color: #e4d2a0;
}

.date {
    position: absolute;
    top: 50px;
    color: black;
}
  
.options {
    flex: 1;
    flex-direction: column;
    background-color: #febb02;
    position: sticky;
    width: fit-content;
    box-shadow: none;
}

.optionItem {
    width: 165px;
    display: flex;
    justify-content: space-between;
    margin: 10px;
}

.optionCounter {
    display: flex;
    align-items: center;
    gap: 10px;
    font-size: 12px;
    color: black;
}

.optionCounterButton {
    width: 20px;
    height: 20px;
    border: 1px solid #0071c2;
    color: #febb02;
    cursor: pointer;
    background-color: #0071c2;
}

.optionCounterButton:disabled {
    cursor: not-allowed;
}