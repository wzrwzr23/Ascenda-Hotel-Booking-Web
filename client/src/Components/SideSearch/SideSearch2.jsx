import Footer from '../../Components/Footer/Footer'
import Header from '../../Components/Header/Header'
import Navbar from '../../Components/Navbar/Navbar'
import React from 'react'
import { useContext, useState } from 'react'
import destData from '../../destinations.json'
import './SideSearch.css'
import { format } from "date-fns";
import { DateRange } from 'react-date-range'
import { SearchContext } from '../../Context/SearchContext'
import { useNavigate } from "react-router-dom";
import { Link } from 'react-router-dom';

// import './DestinationSearch.css'

const SideSearch = () => {
  const [userDest, setUserDest] = useState('')
  const [filteredDest, setFilteredDest] = useState([])
  const [checkInDate, setCheckInDate] = useState(null)
  const [checkOutDate, setCheckOutDate] = useState(null)
  const [numChild, setNumChild] = useState()
  const [numAdult, setNumAdult] = useState()
  const [numRoom, setNumRoom] = useState()
  const [destId, setDestId] = useState()
  const [destination, setDestination] = useState("");
  const [openDate, setOpenDate] = useState(false);
  const [dates, setDates] = useState([
    {
      startDate: new Date(),
      endDate: updateDate(),
      key: "selection",
    },
  ]);
  const [openOptions, setOpenOptions] = useState(false);
  const [options, setOptions] = useState({
    guest: 1,
    // children: 0,
    room: 1,
  });
  const handleOption = (name, operation) => {
    setOptions((prev) => {
      return {
        ...prev,
        [name]: operation === "i" ? options[name] + 1 : options[name] - 1,
      };
    });
  };

  const {dispatch} = useContext(SearchContext)
  const navigate = useNavigate();

  const handleSearch = async () => {
    // await this.timeout(6500)
    const uid = onSearch()
    setDestId("yo", uid);
    console.log(uid)
    const url = `/hotelsearch/${uid}`;
    // const url = "/login"
    dispatch({type:"NEW_SEARCH", payload:{destination, dates, options}})
    navigate(url, { state: { destination, dates, options } });
  };

  const searchDest = (event) => {
    const userSearch = event
    const lowercase = userSearch.toLowerCase()
    setUserDest(userSearch)
    const filterDest = destData.filter((item) => {
      const destTerm = item.term.toLowerCase()
      return lowercase && destTerm.startsWith(lowercase) && destTerm !== lowercase
    })
    if (userSearch === '') {
      setFilteredDest([])
    } else {setFilteredDest(filterDest.slice(0, 10))}
  }
  // function onSearch(userDest) {
  //   if (userDest === '') {
  //     alert('please fill in all input fields.')
  //   } else {
  //     return getUID(userDest, destData)
  //   }
  // }
  // const onSearch = () => {
  //   if (userDest) {
  //     getUID(userDest, destData)
  //   }
  // }

  function updateDate() {
    var date = new Date()
    var currentMonth = date.getMonth();
    var currentDate = date.getDate() + 1;
    var currentYear = date.getFullYear();
    var doo = new Date(currentYear, currentMonth, currentDate)
    return new Date(doo.getTime() + Math.abs(doo.getTimezoneOffset() * 60000))
    // return doo

  }

  function onSearch(){
    var uid;
    var dict = {};
    if (userDest) {
      uid = getUID(userDest, destData)
    }
    
    var checkindate = JSON.stringify(dates[0].startDate).slice(1, 11)
    var checkoutdate = JSON.stringify(dates[0].endDate).slice(1, 11)
    var numguest = options.guest
    var numroom = options.room
    dict['Check-In Date'] = checkindate
    dict['Check-Out Date'] = checkoutdate
    dict['Number of Guests'] = numguest
    dict['Number of Rooms'] = numroom
    console.log("helllllllllo")
    // dispatch({type:"RESET_SEARCH"})
    // console.log("reset")
    dispatch({type:"NEW_SEARCH", payload:{destination, dates, options}})
    console.log("dispatched")
    //alert(JSON.stringify(dict))
    if (uid == null || checkindate == checkoutdate){
      console.log("null");
      navigate("/error");
    }
    else{
      navigate("/hotelsearch2/"+uid, { state: { destination, dates, options } });
    }
    /*navigate("/hotelsearch/"+uid, { state: { destination, dates, options } });*/
  }

  function getUID(value, file) {
    for (var i=0; i<file.length; i++) {
      var obj = file[i]
      if (value === obj.term) {
        setDestId(obj.uid)
        console.log("nami",destId)
        return obj.uid;
      }
    }
    return null;
  }


  return (
    // <div>Destination Search</div>
    <section className='destform'>
        <div className="searchItems">
        <h2>Enjoy Your Stay</h2>  
        <span>Search and Book Hotel</span>
          <div className='usersearch'>
            <input className='searchcity' type='text' placeholder='Search City' name='' id='input' value={userDest} onChange={(e) => {searchDest(e.target.value); setDestination(e.target.value);}} data-testid='searchinput'/>
          </div>
          <div className='dropdown' data-testid="filter-dest">
            {filteredDest.map((item) => {return (
              <div  className='dropdown-row' style={{color: 'black'}} onClick={() => searchDest(item.term)}>{item.term}</div>
            )})}
          </div>
          <div className='lsItem'>
            <span
              onClick={() => setOpenDate(!openDate)}
              className="headerSearchText"
            >{`${format(dates[0].startDate, "MM/dd/yyyy")} to ${format(
                dates[0].endDate,
                "MM/dd/yyyy"
              )}`}</span>
              {openDate && (
                <DateRange
                  /*editableDateInputs={true}*/
                  onChange={(item) => setDates([item.selection])}
                  moveRangeOnFirstSelection={false}
                  ranges={dates}
                  minDate={new Date()}
                  endDatePlaceholder={updateDate()}
                  endDate={updateDate()}
                />
              )}
          </div>
          <div className="options">
            <div className="optionItem">
              <span className="optionText">Guests</span>
              <div className="optionCounter">
                <button
                  disabled={options.guest <= 1}
                  className="optionCounterButton"
                  onClick={() => handleOption("guest", "d")}
                >
                  -
                </button>
                <span className="optionCounterNumber">
                  {options.guest}
                </span>
                <button
                  className="optionCounterButton"
                  onClick={() => handleOption("guest", "i")}
                >
                  +
                </button>
              </div>
            </div>
            {/* <div className="optionItem">
              <span className="optionText">Children</span>
              <div className="optionCounter"> */}
                {/* <button
                  disabled={options.children <= 0}
                  className="optionCounterButton"
                  onClick={() => handleOption("children", "d")}
                >
                  -
                </button> */}
                {/* <span className="optionCounterNumber">
                  {options.children}
                </span>
                <button
                  className="optionCounterButton"
                  onClick={() => handleOption("children", "i")}
                >
                  +
                </button>
              </div>
            </div> */}
            <div className="optionItem">
              <span className="optionText">Room</span>
              <div className="optionCounter">
                <button
                  disabled={options.room <= 1}
                  className="optionCounterButton"
                  onClick={() => handleOption("room", "d")}
                >
                  -
                </button>
                <span className="optionCounterNumber">
                  {options.room}
                </span>
                <button
                  className="optionCounterButton"
                  onClick={() => handleOption("room", "i")}
                >
                  +
                </button>
              </div>
            </div>
          </div>
          {/* <Link to={`/hotelsearch/${destId}`}>
              <button className="siCheckButton" onClick={handleSearch}>See availability</button>
          </Link> */}
          {/* <button onClick={handleSearch}>Search</button> */}
        {/*<button type='Submit' className='submitDest' data-testid='submit'*/}
      {/* <a className='searchBtn' href={'/hotelsearch/'+destId} onClick={onSearch}>Search</a> */}
      <button id="search_button" onClick={onSearch} data-testid='submit'>Search</button>
        {/*</button>*/}
      </div>
    </section>
  ) 
}

export default SideSearch