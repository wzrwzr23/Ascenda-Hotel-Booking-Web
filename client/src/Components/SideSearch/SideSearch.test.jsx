import React from "react";
import SideSearch from "./SideSearch";
import { render, fireEvent, queryByTestId, getByTestId, getByPlaceholderText} from "@testing-library/react"
import { BrowserRouter } from "react-router-dom";
import user from '@testing-library/user-event'
import renderer from "react-test-renderer"



it('renders correctly', () => {
    const {getByTestId, queryByPlaceholderText} = render(<BrowserRouter><SideSearch/></BrowserRouter>);
    expect(getByTestId('submit')).toBeTruthy();
    expect(getByTestId('guestminus')).toBeTruthy();
    expect(getByTestId('guestplus')).toBeTruthy();
    expect(getByTestId('roomminus')).toBeTruthy();
    expect(getByTestId('roomplus')).toBeTruthy();
    expect(getByTestId('userdate')).toBeTruthy();
    expect(queryByPlaceholderText('Search City')).toBeTruthy();
})

describe('Destination search input: Typing', () => {
    test('search box updates on change', () => {
        const {queryByPlaceholderText} = render(<BrowserRouter><SideSearch/></BrowserRouter>);
        const searchInput = queryByPlaceholderText('Search City');
        fireEvent.change(searchInput, {target: {value: "New Delhi"}});
        // userEvent.type(searchInput, 'test')
        expect(searchInput.value).toBe("New Delhi");
    });
})

describe('Guest count (+)', () => {
    test('guest count updates on change', () => {
        const {getByTestId} = render(<BrowserRouter><SideSearch/></BrowserRouter>)
        const guestcount = getByTestId('guestno')
        fireEvent.click(getByTestId('guestplus'))
        expect(guestcount.value).not.toBe(1)
    })
})

describe('Guest count (-)', () => {
    test('guest count does not updates on change', () => {
        const {getByTestId} = render(<BrowserRouter><SideSearch/></BrowserRouter>)
        const guestcount = getByTestId('guestno')
        fireEvent.click(getByTestId('guestminus'))
        expect(guestcount.value).toBe(undefined)
    })
})

describe('Room count (+)', () => {
    test('guest count updates on change', () => {
        const {getByTestId} = render(<BrowserRouter><SideSearch/></BrowserRouter>)
        const guestcount = getByTestId('roomno')
        fireEvent.click(getByTestId('roomplus'))
        expect(guestcount.value).not.toBe(1)
    })
})

describe('Guest count (-)', () => {
    test('guest count does not updates on change', () => {
        const {getByTestId} = render(<BrowserRouter><SideSearch/></BrowserRouter>)
        const guestcount = getByTestId('roomno')
        fireEvent.click(getByTestId('roomminus'))
        expect(guestcount.value).toBe(undefined)
    })
})

it("matches snapshot", ()=>{ //if you change the code in footer, then this test will fail. if intentional, can update snapshot. if by mistake, undo the code
    const tree = renderer.create(<BrowserRouter><SideSearch/></BrowserRouter>).toJSON();
    expect(tree).toMatchSnapshot();
})

// describe('Search button', () => {
//     test('empty destination does not trigger search(to give uid)', () => {
//         jest.spyOn(Alert, 'alert')
//         const {getByTestId, queryByPlaceholderText} = render(<BrowserRouter><DestinationSearch/></BrowserRouter>)
//         fireEvent.click(getByTestId('submit'))
//         expect(Alert.alert).toHaveBeenCalled()
//     })
// })

// describe('Destination Result List: Clicking', () => {
//     test('results on the dropdown list can be clicked', () => {
//         const searchDest = jest.fn()
//         const {getByTestId} = render(<BrowserRouter><DestinationSearch searchDest={searchDest}/></BrowserRouter>);
//         fireEvent.click(getByTestId('filter-dest'));
//         expect(searchDest).toHaveBeenCalled()

//     })
// })

// describe('Destination search input: Empty input', () => {
//     test('does not show any destinations', () => {
//         const onSearch = jest.fn()
//         const {getByTestId} = render(<DestinationSearch/>);
//         fireEvent.change(getByTestId('searchinput'), {target: {value: ""}});
//         expect(onSearch).not.toHaveBeenCalled;
//     })
// })

// describe('Destination search input: Destination clicked', () => {
//     test('search box updates', () => {
//         // const onSearch = jest.fn();
//         const {getByTestId} = render(<DestinationSearch/>);
//         const searchInput = getByTestId('searchinput');
//         fireEvent.click(getByTestId('searchinput'));
//         expect(searchInput.value).toBe();
//     })
// })