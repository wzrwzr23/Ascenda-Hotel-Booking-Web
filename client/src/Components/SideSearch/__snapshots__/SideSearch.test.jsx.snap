// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<section
  className="destform"
>
  <div
    className="searchItems"
  >
    <h2>
      Enjoy Your Stay
    </h2>
    <span>
      Search and Book Hotel
    </span>
    <div
      className="usersearch"
    >
      <input
        className="searchcity"
        data-testid="searchinput"
        id="input"
        name=""
        onChange={[Function]}
        placeholder="Search City"
        type="text"
        value=""
      />
    </div>
    <div
      className="dropdown"
      data-testid="filter-dest"
    />
    <div
      className="lsItem"
      data-testid="guestno"
    >
      <span
        className="headerSearchText"
        data-testid="userdate"
        onClick={[Function]}
      >
        08/10/2022 to 08/11/2022
      </span>
    </div>
    <div
      className="options"
    >
      <div
        className="optionItem"
      >
        <span
          className="optionText"
        >
          Guests
        </span>
        <div
          className="optionCounter"
        >
          <button
            className="optionCounterButton"
            data-testid="guestminus"
            disabled={true}
            onClick={[Function]}
          >
            -
          </button>
          <span
            className="optionCounterNumber"
          >
            1
          </span>
          <button
            className="optionCounterButton"
            data-testid="guestplus"
            onClick={[Function]}
          >
            +
          </button>
        </div>
      </div>
      <div
        className="optionItem"
        data-testid="roomno"
      >
        <span
          className="optionText"
        >
          Room
        </span>
        <div
          className="optionCounter"
        >
          <button
            className="optionCounterButton"
            data-testid="roomminus"
            disabled={true}
            onClick={[Function]}
          >
            -
          </button>
          <span
            className="optionCounterNumber"
          >
            1
          </span>
          <button
            className="optionCounterButton"
            data-testid="roomplus"
            onClick={[Function]}
          >
            +
          </button>
        </div>
      </div>
    </div>
    <button
      data-testid="submit"
      id="search_button"
      onClick={[Function]}
    >
      Search
    </button>
  </div>
</section>
`;
